### PR TITLE
Respect enable_hashagg in the master planner

### PR DESCRIPTION
**Pull request by metdos**
DESCRIPTION: Adds support for respecting enable_hashagg in the master plan

Provides a workaround for #1932

### CODE REVIEW

**Comment by metdos** [comment]
I see weird PostgreSQL version errors on Travis for PostgreSQL 9.6.

```
 -- print whether we're using version > 9 to make version-specific tests clear
 SHOW server_version \gset
 SELECT substring(:'server_version', '\d+')::int > 9 AS version_above_nine;
  version_above_nine 
 --------------------
- t
+ f
```

Any ideas? Maybe @jasonmp85

**Comment by metdos** [comment]
@saicitus, @samay-sharma is there a way to test if this helps with the memory problem on some hll queries?

**Comment by marcocitus** [comment]
> I see weird PostgreSQL version errors on Travis for PostgreSQL 9.6.

The regression tests isn't aware of versions. It just shows the diff with the minimal amount of changes, so it's comparing the 9.6 output to the 10 expected output, but it means that something about the 9.6 output changed.

File: src/backend/distributed/planner/multi_master_planner.c [hunk 1]
```diff
@@ -303,7 +304,7 @@ BuildAggregatePlan(Query *masterQuery, Plan *subPlan)
 		 * see nodeAgg.c:build_pertrans_for_aggref(). In that case we use
 		 * sorted agg strategy, otherwise we use hash strategy.
 		 */
-		if (!groupingIsHashable || hasDistinctAggregate)
+		if (!enable_hashagg || !groupingIsHashable || hasDistinctAggregate)
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_master_planner.c]: @lithp / @mtuncer  will using this GUC cause trouble on windows?

**Comment by metdos** [comment]
> It just shows the diff with the minimal amount of changes

Makes sense.

File: src/backend/distributed/planner/multi_master_planner.c (hunk 1 above)
Comment (mtuncer) [reply src/backend/distributed/planner/multi_master_planner.c]: yes it causes build failure.  I do not think we should commit this change as it is.  
Is there a way to get this value by other means ? (spi etc)

Comment (metdos) [reply src/backend/distributed/planner/multi_master_planner.c]: Hmm, that's interesting. Don't we use similar GUC variables all around the Citus? Do we use only Citus GUC variables?

Comment (mtuncer) [reply src/backend/distributed/planner/multi_master_planner.c]: On second thought,  we can get this checked in. But we have to make sure that enable_hashagg is declared with PGDLLIMPORT. There are other variables declared with this in ```cost.h```

File: src/backend/distributed/planner/multi_master_planner.c [hunk 2]
```diff
@@ -303,7 +304,7 @@ BuildAggregatePlan(Query *masterQuery, Plan *subPlan)
...
 		{
 			if (!groupingIsSortable)
 			{
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: following error message is now incorrect.  
```
if (!enable_hashagg && groupingIsHashable && !hasDistinctAggregate)
``
setting ```enable_hasagg``` may get it to work. 
Perhaps we could provide this as hint.

Comment (metdos) [reply src/backend/distributed/planner/multi_master_planner.c]: But the default for `enable_hasagg` is already on.

Comment (mtuncer) [reply src/backend/distributed/planner/multi_master_planner.c]: I meant hint could be "setting enable_hashagg to true may help ...."

File: src/backend/distributed/planner/multi_master_planner.c [hunk 3]
```diff
@@ -30,6 +30,9 @@
 #include "utils/syscache.h"
 
 
+extern PGDLLIMPORT bool enable_hashagg;
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: I think the right fix is to make this inside postgresql. We should open an issue to track this.

Comment (metdos) [reply src/backend/distributed/planner/multi_master_planner.c]: 1. We can keep this temporarily until PostgreSQL gets it. Does that work?

File: src/backend/distributed/planner/multi_master_planner.c [hunk 4]
```diff
@@ -303,13 +304,15 @@ BuildAggregatePlan(Query *masterQuery, Plan *subPlan)
...
-		if (!groupingIsHashable || hasDistinctAggregate)
+		if (!enable_hashagg || !groupingIsHashable || hasDistinctAggregate)
 		{
 			if (!groupingIsSortable)
 			{
 				ereport(ERROR, (errmsg("grouped column list must cannot be sorted"),
 								errdetail("Having a distinct aggregate requires "
-										  "grouped column list to be sortable.")));
+										  "grouped column list to be sortable."),
+								errhint("Setting enable_hashagg to on may "
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: we should show this hint only if ```enable_hashagg``` is not set

Comment (mtuncer) [reply src/backend/distributed/planner/multi_master_planner.c]: try
```c
enable_hashagg?NULL:errhint(...)
```

**Comment by metdos** [comment]
> using enable_hasagg breaks windows builds. We can live with it at the moment. But we should get it fixed, or provide a work around

Sent a patch to PostgreSQL

https://www.postgresql.org/message-id/CAL1dPcfa45o1dC-c4t-48v0OZE6oy4ChJhObrtkK8mzNfXqDTA%40mail.gmail.com

File: src/backend/distributed/planner/multi_master_planner.c [hunk 5]
```diff
@@ -303,13 +304,21 @@ BuildAggregatePlan(Query *masterQuery, Plan *subPlan)
...
 		 * sorted agg strategy, otherwise we use hash strategy.
 		 */
-		if (!groupingIsHashable || hasDistinctAggregate)
+		if (!enable_hashagg || !groupingIsHashable || hasDistinctAggregate)
 		{
+			char *messageHint = NULL;
+			if (!enable_hashagg && groupingIsHashable)
+			{
+				messageHint = "Setting enable_hashagg to on may help "
+							  "if grouped column list is hashable.";
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: ```groupingIsHashable``` asserts that grouping here is hashable. We do not need to second hint line. Perhaps something like
```
"Consider setting enable_hashagg to on."
```

**Comment by metdos** [comment]
The required change is merged into PostgreSQL master branch.
https://github.com/postgres/postgres/commit/935dee9ad5a8d12f4d3b772a6e6c99d245e5ad44

**Comment by lithp** [comment]
@metdos I think that we should ask for this commit to be backported to at least `REL_10_STABLE`. It's currently in master, which means we won't run on Windows until version postgres 11 is realeased.

**Comment by jmehnle** [comment]
@metdos, @ozgune (since you've handled technical issues with us in the past): Is there any chance this workaround can be backported to Citus 6.2?

**Pull request merged**